bpi-r4: init

newAM opened the pull request:
Thanks for this repository, it is a great reference!

I got the bpi-r4 booting here: <https://github.com/newAM/nixos-bpi-r4>
I hope to bring that code back into this repository.

Notes:
* No wireless yet.  I do not have a wireless module to test with.
* This PR is untested, it will take a while for my computer to compile this all natively.

newAM committed:
bpi-r4: init

tbaumann commented:
Hah, funny. I was just about to start that task. Literally started a few hours ago. But I don't have the HW yet.

tbaumann commented:
Do you need to use the custom kernel? Upstream should have support.

newAM commented:
I'm glad I opened a draft then :)

An upstream kernel should work.  I'll investigate that.  Looking back at my notes it was failing to find the EXT4 root partition with an upstream kernel, but it is likely the underlying cause is a bug in the SD card boot device tree overlay, not anything to do with the kernel itself.

newAM commented:
I tried out an upstream kernel and the SD card boot device tree overlay does not apply anymore.

```text
linux-aarch64-unknown-linux-gnu-6.10.2-modules-shrunk> building '/nix/store/8yjqnwnrv1v6h0ic1vgvnxlg2zzxmsag-linux-aarch64-unknown-linux-gnu-6.10.2-modules-shrunk.drv'
device-tree-overlays> ./mediatek -> /nix/store/7gshz5sm2y35a4kmvj6cyhz6icb8z22z-device-tree-overlays/./mediatek
device-tree-overlays> './mediatek/mt7988a-bananapi-bpi-r4.dtb' -> '/nix/store/7gshz5sm2y35a4kmvj6cyhz6icb8z22z-device-tree-overlays/./mediatek/mt7988a-bananapi-bpi-r4.dtb'
device-tree-overlays> Applying overlay BananaPi BPiR4 Enable SD card interface to mt7988a-bananapi-bpi-r4.dtb... 
device-tree-overlays> Failed to apply '/nix/store/rlg3kizk7z2pm692qlw9zvrab12bjl2c-BananaPi-BPiR4-Enable-SD-card-interface-dtbo-aarch64-unknown-linux-gnu': FDT_ERR_NOTFOUND
```

Any ideas?

nakato commented on pkgs/bananaPiR4/mt7988-nixos.env:
```diff
@@ -0,0 +1,16 @@
+// No idea if these are correct.
+// Using values from the R3 will result in an unbootable system.
+// I messed around until it booted.
```
Does it boot without these values being manually specified?

They exist on the R3 as the R3 fails to boot without setting them.  I think it was a combination of being set for small kernels/initrds and reserved memory locations.

Upstream doesn't mention reserved memory, but it looks like *0x43000000* for *0x50000* is reserved.

newAM replied:
I tried without the environment file and got this error at boot: `missing environment variable: kernel_addr_r`

Full log: https://github.com/newAM/nixos-bpi-r4/blob/a13c878b4a6c8a894c0f43ade0e9986ebc94d493/bootlog.txt

Edit: There's probably a small subset of these needed.  I'll try removing some of them.

newAM replied:
I was able to minify it to just 3 values, `fdt_addr_r`, `kernel_addr_r`, `ramdisk_addr_r`.  None of the others were required to boot.

nakato commented on pkgs/bananaPiR4/mt7988-nixos.env:
```diff
@@ -0,0 +1,16 @@
+ramdisk_addr_r=0x50000000
+scriptaddr=0x90000000
+
+// Set initrd high to be under the reserved memory
```
This comment doesn't make since with the lack of reserved memory being avoided.

newAM replied:
Done!

nakato commented on pkgs/bananaPiR4/default.nix:
```diff
@@ -0,0 +1,81 @@
+    });
+
+
+  linuxPackages_6_9_bananaPiR4 = linuxKernel.packagesFor (linux_6_9.override {
```
This needs to be renamed in a way to make it clear that this kernel is not an upstream kernel.

Can we also add something like

```
extraMeta.vendorKernel = true;
```
(I know Frank-W isn't a vendor, but I can't think of a better phrasing for "not-upstream")

and use that to throw a warning that users will see during `nixos-rebuild` runs if the vendor kernel is not explicitly set by the user.


That warning should say something along the lines of:

> This device uses a vendor kernel, vendor kernels commonly use different names for various nodes in the device-tree.  If you are not manually applying device-tree overlays or somehow relying on DT node names in userspace, which is unlikely, you can ignore this warning.  If you are, explicitly set `boot.kernelPackages` to the vendor kernel.  Note that this kernel will be dropped when upstream support for the kernel becomes available.

nakato commented on modules/boards/bananapi/bpir4/default.nix:
```diff
@@ -0,0 +1,57 @@
+      model = "bpir4";
+      dtRoot = "mediatek,mt7988a";
+
+      i2c.devices.i2c0 = {
+        status = "okay";
+      };
```
Can you add
```
i2c.devices.i2c1 = {
  status = "disabled";
  enableMethod.dtOverlay = {
    enable = true;
    pinctrl-names = "default";
    # Pins 3, 5
    pinctrl-0 = lib.mkDefault ["i2c1_pins"];
  };
};
```

and
```
i2c.devices.i2c2 = {
  status = "okay";
}
```

newAM replied:
Done!

nakato commented on modules/boards/bananapi/bpir4/default.nix:
```diff
@@ -0,0 +1,57 @@
+        deviceName = "ttyS0";
+        console = true;
+      };
+
```
```suggestion
      # uart.devices.uart1 not currently in DT.  GPIO header pins 11, 13.

```

newAM replied:
Done!

nakato commented on modules/boards/bananapi/bpir4/default.nix:
```diff
@@ -0,0 +1,57 @@
+        console = true;
+      };
+
+      wifi.devices.wifi = {
+        status = "okay";
+        enableMethod.moduleLoad = ["mt7915e"];
+        disableMethod.dtOverlay.enable = true;
+        enable = lib.mkDefault config.sbc.wireless.wifi.enable;
+      };
```
WiFi on the R4 is on PCIe, but still needs some DT setup, so this will have to change, but I'm not sure how yet.  It will probably not be included at `sbc.board.wifi` as this is all DeviceTree parts.   We have `config.sbc.wireless.wifi.enable`, so the configuration could be done in `./bpir4/wifi.nix` under a `config = lib.mkIf config.sbc.wireless.wifi.enable { };`

Looks like Frank-W enables i2c on the miniPCIe ports and configures the eeproms in a way the wifi driver knows to check.

We should also probably have something that, on boot, checks those EEPROM's have a MAC address, and that it is valid, that it isn't all 0's, and isn't an default burned into all devices during production, and if it is, burn a new randomly generated one into the eeproms before the wifi kernel modules load.

nakato replied:
I just noticed you said you didn't get the wifi module.

We can drop this section for now then and wifi support can be added in the future.

newAM replied:
Sounds good!  Dropped.

tbaumann replied:
WiFi is PICIe anyway. There are I think a few modules sold by Sinovip, but I believe you can use what you like.

But perhaps the module for https://wiki.banana-pi.org/Banana_Pi_BPI-R4#WiFi7_iPA_NIC_Module_support should be included. Those are the default hardware. 

There are two patches linked on that page...

They started shipping those boards now. I'm getting one...

nakato replied:
While it is PCIe, that module does require additional setup to work appropriately from what I can tell.

At the very least, from what I can see, the MAC addresses comes from EEPROMs and those are connected over an i2c bus back to the CPU, so some handling is required.  I don't know if they are all zero, or if they're burned with unique, or identical macs.  It's important to know what the driver will do with those if the EEPROM read fails, or the contents are invalid.  Depending on the driver we might be able to handle the wireless MACs via the device-tree similar to the R3 but I don't know what module it even loads right now.
I suspect we'll need to replicate the MAC setting logic from the R3 for the ethernet interfaces at least.  But that will be a matter of question after basic functionality is working.

While it is up to the wireless hardware operator to ensure they are operating withing the law of their region, any config we do needs to make sane choices that make it hard to do the wrong thing.

nakato commented on modules/boards/bananapi/bpir4/default.nix:
```diff
@@ -0,0 +1,57 @@
+
+    sbc.board = {
+      vendor = "BananaPi";
+      model = "bpir4";
```
BPiR4

newAM replied:
Done!

nakato commented:
> I tried out an upstream kernel and the SD card boot device tree overlay does not apply anymore.
> 
> Any ideas?

The upstream device-tree for the device looks to be not much more than a placeholder.  You might be able to create a complete dts for the device similar to what I've done with the [OrangePi 5B](https://github.com/nakato/nixos-sbc/blob/ddb8f1a16f1b6ebb32cb7b13125e45d65673905e/pkgs/orangepi5/default.nix#L40) and use that with an upstream kernel provided the hardware is supported upstream, or more likely the R3 kernel with a few additional options set as I suspect the DSA driver is still buggy when built as a module rather than built in.  I would not be opposed to the R3 kernel supporting R3 and R4 if that were the case.


As much as I do not want non-upstream kernels in here, for now it does look like the only option as the device is in the early days and not everything has hit the mailing list yet, which would allow building upstream with the patches from lore/patchwork.

newAM commented:
I have been tinkering with the kernel, I got it to boot an upstream kernel by building the DTB separately as you did for the OrangePi 5.

Sadly the upstream kernel panics.  Looking through Frank-W's changes I think there are modules needed that are not upstream, specifically the GPIO driver seems pretty important, and some clock changes.

I'm going to tinker with the kernel some more.  I'm not too keen on vendor kernels either.  I would prefer to have a set of patches ontop of a release.  I think this is doable with a dozen or so patches.

tbaumann commented:
For uboot env, this could be a template https://github.com/BPI-SINOVOIP/BPI-R4-bsp-6.1/blob/main/u-boot-mt/bananapi_bpi-r4_sdmmc_env

nakato commented:
> For uboot env, this could be a template https://github.com/BPI-SINOVOIP/BPI-R4-bsp-6.1/blob/main/u-boot-mt/bananapi_bpi-r4_sdmmc_env

No, only the newer bootflow method is used in this repo.  This keeps boot consistent across boards and keeps the u-boot builds maintainable by me.

nakato commented:
> I'm going to tinker with the kernel some more. I'm not too keen on vendor kernels either. I would prefer to have a set of patches ontop of a release. I think this is doable with a dozen or so patches.

I have been working on bringing down the kernel compile time to something reasonable without needing a beefy build box.  My goal is to get these building on ARM in CI and pushed into a cache (on my second pass at that now).

I just got around to testing the [linuxPackages_6_9_bananaPiR3_minimal](https://github.com/nakato/nixos-sbc/blob/504536864204d6a224b5453800f123cb6c00f2cc/pkgs/bananaPiR3/default.nix#L148) image I've been using in my CI tests on my R3 that I use as an AP and it works, so it should be good to go as a base if you want quicker builds of the kernel you're working on with patches. 

I forgot to time the build, but I think it was under an hour on my OrangePi 5B (with the OPi5B being about 3x faster than a RPi4).

Though if you're rebuilding a lot, it's probably easier to work in a devshell and leverage incremental builds.  That's what I did when I was getting the first kernel ready for the R3 with a ton of patches pulled from the mailing lists.

newAM commented:
### Kernel

I tinkered with the kernel a lot, and in the end Frank W's is the most reasonable way forward.

Omitting some of Frank's commits that don't seem important often leads to subtle breakage.

I can just apply all ~20ish of Franks commits as patches, but not all the patches apply since the `nixpkgs` kernel is newer than Franks (for example Frank's 6.10 is based on 6.10.0, whereas nixpkgs is on 6.10.2).

Many patches Frank's commits are patches from the mailing list, but patches submitted in January & February have only just started making it into a release (6.11-rc).  By my estimate we will be able to use an upstream kernel hopefully around 6.13.

I updated the kernel to 6.10 for the BPi since it's _less_ patched than Frank's 6.9.  Additionally 6.10 is still supported upstream.  This required a flake.lock update.  I can move the flake.lock update to a separate commit / pull-request if desired.

> I just got around to testing the [linuxPackages_6_9_bananaPiR3_minimal](https://github.com/nakato/nixos-sbc/blob/504536864204d6a224b5453800f123cb6c00f2cc/pkgs/bananaPiR3/default.nix#L148) image I've been using in my CI tests on my R3 that I use as an AP and it works, so it should be good to go as a base if you want quicker builds of the kernel you're working on with patches.

That looks great!  I'll try that out later.

## Testing

I have not tested this yet, which is why it's still a draft.

I'm building it all now, if it boots I'll mark this ready for review, but I don't expect any more code changes (apart from review feedback).

## Other Notes

Often the `mtk_soc` module will fail to load during boot.  A simple `sudo rmmod mtk_eth` /  `sudo modprobe mtk_eth` will fix this.

<details>
<summary>Example logs...</summary>

```console
$ sudo rmmod mtk_eth
$ sudo modprobe mtk_eth
[   76.274207] mtk_soc_eth 15100000.ethernet: generated random MAC address 20:08:02:00:00:00
[   76.282419] mtk_soc_eth 15100000.ethernet: generated random MAC address 20:08:02:00:00:00
[   76.290680] mtk_soc_eth 15100000.ethernet: generated random MAC address 20:08:02:00:00:00
[   76.308182] mtk_soc_eth 15100000.ethernet eth0: mediatek frame engine at 0xffffffc083a00000, irq 112
[   76.318370] mtk_soc_eth 15100000.ethernet eth1: mediatek frame engine at 0xffffffc083a00000, irq 112
[   76.328851] mtk_soc_eth 15100000.ethernet eth2: mediatek frame engine at 0xffffffc083a00000, irq 112
[   76.473227] mt7530-mmio 15020000.switch: configuring for fixed/internal link mode
[   76.480876] mt7530-mmio 15020000.switch: Link is Up - 10Gbps/Full - flow control rx/tx
[   76.506101] mt7530-mmio 15020000.switch wan (uninitialized): PHY [mt7530-0:00] driver [MediaTek MT7988 PHY] (irq=POLL)
[   76.543193] mt7530-mmio 15020000.switch lan1 (uninitialized): PHY [mt7530-0:01] driver [MediaTek MT7988 PHY] (irq=POLL)
[   76.580563] mt7530-mmio 15020000.switch lan2 (uninitialized): PHY [mt7530-0:02] driver [MediaTek MT7988 PHY] (irq=POLL)
[   76.620298] mt7530-mmio 15020000.switch lan3 (uninitialized): PHY [mt7530-0:03] driver [MediaTek MT7988 PHY] (irq=POLL)
[   76.631916] mtk_soc_eth 15100000.ethernet eth0: entered promiscuous mode
[   76.638727] DSA: tree 0 setup
[   76.683770] br-lan: port 1(lan2) entered blocking state
[   76.689012] br-lan: port 1(lan2) entered disabled state
[   76.694326] mt7530-mmio 15020000.switch lan2: entered allmulticast mode
[   76.700943] mtk_soc_eth 15100000.ethernet eth0: entered allmulticast mode
[   76.707988] mt7530-mmio 15020000.switch lan2: entered promiscuous mode
[   76.721687] mtk_soc_eth 15100000.ethernet eth0: configuring for fixed/internal link mode
[   76.729865] mtk_soc_eth 15100000.ethernet eth0: Link is Up - 10Gbps/Full - flow control rx/tx
[   76.730559] mt7530-mmio 15020000.switch wan: configuring for phy/internal link mode
[   76.748133] mt7530-mmio 15020000.switch wan: Link is Up - 1Gbps/Full - flow control rx/tx
[   76.749271] br-lan: port 2(lan3) entered blocking state
[   76.761570] br-lan: port 2(lan3) entered disabled state
[   76.766855] mt7530-mmio 15020000.switch lan3: entered allmulticast mode
[   76.773722] mt7530-mmio 15020000.switch lan3: entered promiscuous mode
[   76.864853] mt7530-mmio 15020000.switch lan2: configuring for phy/internal link mode
[   76.874324] mt7530-mmio 15020000.switch lan3: configuring for phy/internal link mode
```
</details>

Sometimes the SD card wont resize itself.  I tried the script from this repository and the original script from nixpkgs, I had intermittent issues regardless of the script.  I don't think this is a code issue with the boot script.  I think this is a race condition with the order of modules in the BPi-R4 bootup.  Or maybe my SD card is just bad.  Still gathering data.  I haven't put any effort into debugging because it wont impact subsequent boots after it is fixed.

Worked around with `e2fsck`.

<details>
<summary>Example logs...</summary>

```console
$ sudo e2fsck -fy /dev/disk/by-uuid/0b5e3376-c7e9-4284-9514-9c3b51244f19
e2fsck 1.47.1 (20-May-2024)
Resize inode not valid.  Recreate? yes

Pass 1: Checking inodes, blocks, and sizes
Pass 2: Checking directory structure
Pass 3: Checking directory connectivity
Pass 3A: Optimizing directories
Pass 4: Checking reference counts
Pass 5: Checking group summary information
Free blocks count wrong for group #0 (4294967295, counted=0).
Fix? yes

Free blocks count wrong (20616, counted=20617).
Fix? yes


root: ***** FILE SYSTEM WAS MODIFIED *****
root: 138244/150784 files (0.1% non-contiguous), 601975/622592 blocks
$ sudo resize2fs /dev/disk/by-uuid/0b5e3376-c7e9-4284-9514-9c3b51244f19
resize2fs 1.47.1 (20-May-2024)
Resizing the filesystem on /dev/disk/by-uuid/0b5e3376-c7e9-4284-9514-9c3b51244f19 to 31291259 (4k) blocks.
The filesystem on /dev/disk/by-uuid/0b5e3376-c7e9-4284-9514-9c3b51244f19 is now 31291259 (4k) blocks long.

```
</details>

newAM commented:
Alright everything boots now.  I did a few more changes, rebased onto the latest `main`, removed some dirty code that I didn't intend to commit, and fixed a typo in the environment file.

nakato commented:
This commit has been merged as b157b70, github does not recognise this, so this is "close with comment".

I pulled out the flake.lock update as it's easier for me to bump them than audit hashes, though in this case they were identical, and also reworded the commit message as I like those to be very verbose.


I've open #11 to track next steps from here.
